ml-algorithms/src/main/java/org/opensearch/ml/engine/processor/MLForEachProcessor.java

```diff
@@ -0,0 +1,217 @@
+        }
+        String pathValue = (String) config.get("path");
+        if (pathValue == null || pathValue.trim().isEmpty()) {
+            throw new IllegalArgumentException("'path' cannot be empty for for_each processor");
+        }
+
+        if (!config.containsKey("processors")) {
+            throw new IllegalArgumentException("'processors' is required for for_each processor");
+        }
+    }
+
+    /**
+     * Processes the input by iterating through the array at the specified path and applying
+     * the configured processors to each element.
+     * <p>
+     * The input is first converted to a JSON string, then parsed as a JSON document.
+     * The array at the specified path is read, and each element is processed through the
+     * processor chain. The modified array is then set back at the original path.
+     * </p>
+     * 
+     * @param input the input object to process (can be a Map, String, or any JSON-serializable object)
+     * @return the modified JSON document with processed array elements, or the original input
+     *         if an error occurs or the path doesn't point to an array
+     */
+    @Override
+    public Object process(Object input) {
+        try {
+            String jsonStr = StringUtils.toJson(input);
+            DocumentContext context = JsonPath.parse(jsonStr);
+
+            // Read the array
+            Object arrayObj;
+            try {
+                arrayObj = context.read(path);
+            } catch (Exception e) {
+                log.warn("Path '{}' does not exist in input: {}", path, e.getMessage());
+                return input;
+            }
+
+            // Validate it's an array
+            if (!(arrayObj instanceof List)) {
+                log.warn("Path '{}' does not point to an array, found type: {}", path, arrayObj.getClass().getSimpleName());
+                return input;
+            }
+
+            List<?> array = (List<?>) arrayObj;
+            if (array.isEmpty()) {
+                log.debug("Array at path '{}' is empty, nothing to process", path);
+                return input;
+            }
+
+            // Process each element
+            List<Object> processedArray = new ArrayList<>();
+            for (int i = 0; i < array.size(); i++) {
+                Object element = array.get(i);
+                try {
+                    Object processed = ProcessorChain.applyProcessors(element, processors);
+                    processedArray.add(processed);
+                } catch (Exception e) {
+                    log.warn("Failed to process element at index {} in path '{}': {}", i, path, e.getMessage());
+                    // Keep original element on error
+                    processedArray.add(element);
+                }
+            }
+
+            // Set the processed array back
+            // Remove [*] wildcards from path for setting
+            String setPath = path.replaceAll("\\[\\*\\]$", "");
```

Comment:
will this work if the path has [*] in the middle? for example, text_chunks[*].text_chunk. it seems this regex pattern might only takes the trailing [*]. 

do you want to remove all [*] in the path? maybe we can do  
`String setPath = path.replaceAll("\\[\\*\\]", "");`

Reply:
oh I saw your test cases, testExtractFactsFromConversation_WithProcessorChainExtractingJson which is parsing `$.content[0].text`, should be fine. approved